kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java

```diff
@@ -0,0 +1,489 @@
+
+public class MongoWindowedTable implements RemoteWindowedTable<WriteModel<WindowDoc>> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MongoWindowedTable.class);
+  private static final String METADATA_COLLECTION_SUFFIX = "_md";
+
+  private static final UpdateOptions UPSERT_OPTIONS = new UpdateOptions().upsert(true);
+
+  private final String name;
+  private final SegmentPartitioner partitioner;
+
+  private final MongoDatabase database;
+  private final MongoCollection<WindowMetadataDoc> metadata;
+  private final ConcurrentMap<Integer, PartitionSegments> kafkaPartitionToSegments =
+      new ConcurrentHashMap<>();
+
+  private class PartitionSegments {
+    private final long epoch;
+    private final SegmentBatch segmentBatch;
+
+    // Recommended to keep the total number of collections under 10,000, so we should not
+    // let num_segments * num_kafka_partitions exceed 10k at the most
+    private final Map<SegmentPartition, MongoCollection<WindowDoc>> segmentWindows;
+
+    public PartitionSegments(
+        final long streamTime,
+        final long epoch
+    ) {
+      this.segmentBatch = new SegmentBatch(streamTime);
+      this.epoch = epoch;
+      this.segmentWindows = new ConcurrentHashMap<>();
+    }
+
+    void createSegment(final SegmentPartition segmentToCreate) {
+      LOG.info("{}[{}] Creating segment id {}",
+               name, segmentToCreate.tablePartition, segmentToCreate.segmentId);
+
+      final MongoCollection<WindowDoc> windowDocs =
+          database.getCollection(collectionNameForSegment(segmentToCreate), WindowDoc.class);
+      windowDocs.createIndex(
+          Indexes.ascending(WindowDoc.WINDOW_START_TS)
+      );
+
+      segmentWindows.put(segmentToCreate, windowDocs);
+    }
+
+    // Note: it is always safe to drop a segment, even without validating the epoch, since we only
+    // ever attempt to expire segments after successfully flushing the batch which rolled them,
+    // thus guaranteeing that even if we are fenced, the new writer will pick up from an offset and
+    // stream-time such that the segment is already expired
+    // However this does leave us vulnerable to zombie segments. We should consider doing a sweep
+    // for old segments every so often with #listCollectionNames, since the segmentId is embedded in
+    // the collection name and will indicate whether it's part of the active set or not, perhaps
+    // using a scheduled executor to avoid blocking processing
+    // In fact we may want to move all the segment expiration to a background process since there's
+    // no async way to drop a collection but the stream thread doesn't actually need to wait for it
+    void expireSegment(final SegmentPartition segmentToDelete) {
+      LOG.info("{}[{}] Expiring segment id {}",
+               name, segmentToDelete.tablePartition, segmentToDelete.segmentId);
+
+      final var expiredDocs = segmentWindows.get(segmentToDelete);
+      expiredDocs.drop();
```

Comment:
what happens if this is called on a collection that doesn't exist? (it's a possible race condition if this thread is a zombie)

Reply:
Great question -- if only there were remotely useful javadocs for Mongo's java API 😞

That said, based on my own testing, it seems to be a no-op -- so this is safe